presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeJoinsOverMapColumns.java

```diff
@@ -0,0 +1,455 @@
+
+            for (VariableReferenceExpression var : joinOutputCols) {
+                assignments.put(var, var);
+            }
+            for (VariableReferenceExpression var : joinKeys) {
+                VariableReferenceExpression newVar = joinKeyMap.get(var);
+
+                VariableReferenceExpression isNullVar = variableAllocator.newVariable(var.getName() + "_NULL", BOOLEAN);
+                assignments.put(isNullVar, specialForm(IS_NULL, BOOLEAN, ImmutableList.of(var)));
+
+                // construct predicate of the form "coalesce(newVar, 0) = coalesce(var, 0)"
+                RowExpression coalesceComp = equalityPredicate(functionResolution, coalesceToZero(newVar), coalesceToZero(var));
+                RowExpression nullComp = equalityPredicate(functionResolution, specialForm(IS_NULL, BOOLEAN, ImmutableList.of(newVar)), isNullVar);
+                nullComparisonBuilder.add(nullComp);
+                coalesceComparisonBuilder.add(coalesceComp);
+            }
+
+            ProjectNode projectNode = new ProjectNode(planNodeIdAllocator.getNextId(), joinNode, assignments.build());
+            List<VariableReferenceExpression> resultOutputCols = Stream.concat(planNode.getOutputVariables().stream(), projectNode.getOutputVariables().stream()).collect(toImmutableList());
+
+            List<RowExpression> joinCriteria = Stream.concat(nullComparisonBuilder.build().stream(), coalesceComparisonBuilder.build().stream()).collect(toImmutableList());
+            JoinNode newJoinNode = new JoinNode(
+                    joinNode.getSourceLocation(),
+                    planNodeIdAllocator.getNextId(),
+                    JoinNode.Type.LEFT,
+                    planNode,
+                    projectNode,
+                    ImmutableList.of(),
+                    resultOutputCols,
+                    Optional.of(LogicalRowExpressions.and(joinCriteria)),
+                    joinNode.getLeftHashVariable(),
+                    joinNode.getRightHashVariable(),
+                    joinNode.getDistributionType(),
+                    joinNode.getDynamicFilters());
+
+            return newJoinNode;
+        }
+
+        boolean validateProjectAssignments(ProjectNode projectNode)
+        {
+            Assignments assignments = projectNode.getAssignments();
+            PlanNode input = projectNode.getSource();
+            ImmutableSet<VariableReferenceExpression> inputColsSet = ImmutableSet.copyOf(input.getOutputVariables());
+
+            for (Map.Entry<VariableReferenceExpression, RowExpression> assignment : assignments.entrySet()) {
+                RowExpression expr = assignment.getValue();
+                if (!inputColsSet.containsAll(VariablesExtractor.extractUnique(expr))) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private RowExpression coalesceToZero(RowExpression var)
+        {
+            RowExpression zero = constant(0L, BIGINT);
```

Comment:
I know we have added the type validation on top, but would be good to check the rowExpression type here as well.